src/Compiler/Checking/NameResolution.fs

```diff
@@ -530,14 +530,18 @@ let IsMethInfoPlainCSharpStyleExtensionMember g m isEnclExtTy (minfo: MethInfo)
 /// Get the info for all the .NET-style extension members listed as static members in the type.
 let private GetCSharpStyleIndexedExtensionMembersForTyconRef (amap: Import.ImportMap) m  (tcrefOfStaticClass: TyconRef) =
     let g = amap.g
-
-    if IsTyconRefUsedForCSharpStyleExtensionMembers g m tcrefOfStaticClass then
-        let pri = NextExtensionMethodPriority()
+    
+    if g.langVersion.SupportsFeature(LanguageFeature.CSharpExtensionAttributeNotRequired) then
         let ty = generalizedTyconRef g tcrefOfStaticClass
+      
+        let minfos =
+            GetImmediateIntrinsicMethInfosOfType (None, AccessorDomain.AccessibleFromSomeFSharpCode) g amap m ty
+            |> List.filter (IsMethInfoPlainCSharpStyleExtensionMember g m true)
```

Comment:
I'm concerned about the performance implications of this preview feature. If my late-night analysis is correct it seems to me we're now looking into the methods of every type in every opened namespace.  This must be traversing a **lot** more metadata than previously.  This will be very significant both from a perf question - think what happens if there are types in referenced assemblies that were previously unused that contain 100,000 methods. It also has some correctness implications see #15158.

Given the feature being implemented this is a little tricky to avoid. We really have to be looking for the `ExtensionAttribute` that mark the types **before** we traverse them, and avoiding the traversal for types that do not contain extension methods.  That's really what the attributes are for.

Now, these attributes will be correctly present in compiled reference assemblies because they are either explicit or added in CheckDeclarations, so for referenced assemblies we can just look at them.  However these attributes are not necessarily present yet in the assembly being compiled - e.g. in the case of `namespace rec` where one part of the recursive set of types opens another before the attributes have been fully stitched in. 

My (late night dont-take-as-gospel) recommendation is that this code be adjusted to look for `ExtensionAttribute`, except in the assembly being compiled.  So something like:

```
    if g.langVersion.SupportsFeature(LanguageFeature.CSharpExtensionAttributeNotRequired) then
        let ty = generalizedTyconRef g tcrefOfStaticClass
        let csharpStyleExtensionMembers = 
            if IsTyconRefUsedForCSharpStyleExtensionMembers g m tcrefOfStaticClass || tcrefOfStaticClass.IsLocalRef then
                GetImmediateIntrinsicMethInfosOfType (None, AccessorDomain.AccessibleFromSomeFSharpCode) g amap m ty
                |> List.filter (IsMethInfoPlainCSharpStyleExtensionMember g m true)
            else
                []
                
        if not minfos.IsEmpty then
                let pri = NextExtensionMethodPriority()
```